File: web/JS/Registro.js
/* 
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/JSP_Servlet/JavaScript.js to edit this template
 */

function verificarCamposVacios() {
    let cedula = document.getElementById("InputCedula");
    let nombre = document.getElementById("InputNombre");
    let apellido = document.getElementById("InputApellido");
    let telefono = document.getElementById("InputTelefono");
    let correo = document.getElementById("InputCorreo");
    let contrasenia = document.getElementById("InputContrasenia");
    let contrasenia2 = document.getElementById("InputContrasenia2");

    let valorCedula = cedula.value;
    let valorNombre = nombre.value;
    let valorApellido = apellido.value;
    let valorTelefono = telefono.value;
    let valorCorreo = correo.value;
    let valorContrasenia = contrasenia.value;
    let valorContrasenia2 = contrasenia2.value;

    let errorCedula = document.getElementById("errorCedula");
    let errorNombre = document.getElementById("errorNombre");
    let errorApellido = document.getElementById("errorApellido");
    let errorTelefono = document.getElementById("errorTelefono");
    let errorCorreo = document.getElementById("errorCorreo");
    let errorContrasenia = document.getElementById("errorContrasenia");
    let errorContrasenia2 = document.getElementById("errorContrasenia2");
    let errorCedulaLong = document.getElementById("errorCedulaLong");
    let errorContraLong = document.getElementById("errorContraLong");
    let errorPwdDesigual = document.getElementById("errorPwdDesigual");

    errorCedulaLong.classList.add('oculto');
    errorContraLong.classList.add('oculto');
    errorCedula.classList.add('oculto');
    errorNombre.classList.add('oculto');
    errorApellido.classList.add('oculto');
    errorCorreo.classList.add('oculto');
    errorTelefono.classList.add('oculto');
    errorContrasenia.classList.add('oculto');
    errorContrasenia2.classList.add('oculto');

    if (valorCorreo === "") {
        correo.classList.add('input-error');
        errorCorreo.classList.remove('oculto');

        return false;
    }

    if (valorCedula === "") {
        cedula.classList.add('input-error');
        errorCedula.classList.remove('oculto');

        return false;
    }

    if (!/^\d{8}$/.test(valorCedula)) {
        cedula.classList.add('input-error');
        errorCedulaLong.classList.remove('oculto');

        return false;
    }

    if (valorNombre === "") {
        nombre.classList.add('input-error');
        errorNombre.classList.remove('oculto');

        return false;
    }

    if (valorApellido === "") {
        apellido.classList.add('input-error');
        errorApellido.classList.remove('oculto');

        return false;
    }

    if (valorTelefono === "") {
        telefono.classList.add('input-error');
        errorTelefono.classList.remove('oculto');

        return false;
    }

    if (!(/^\d{8}$/.test(valorTelefono) || /^\d{9}$/.test(valorTelefono))   ) {
        telefono.classList.add('input-error');
        errorTelefonoLong.classList.remove('oculto');

        return false;
    }

    if (valorContrasenia === "") {
        contrasenia.classList.add('input-error');
        errorContrasenia.classList.remove('oculto');

        return false;
    }

    if (valorContrasenia2 === "") {
        contrasenia2.classList.add('input-error');
        errorContrasenia2.classList.remove('oculto');

        return false;
    }

    if (valorContrasenia.length < 8) {
        contrasenia.classList.add('input-error');
        errorContraLong.classList.remove('oculto');

        return false;
    }

    if (valorContrasenia !== valorContrasenia2) {
        errorPwdDesigual.classList.remove('oculto');
        return false;
    }
    return true;
}


function validarCedula(input) {
    input.value = input.value.replace(/[^0-9]/g, "");

    if (input.value.length > 8) {
        input.value = input.value.slice(0, 8);
    }
}

function clearErrors() {
    let errorCedula = document.getElementById("errorCedula");
    let errorCedulaLong = document.getElementById("errorCedulaLong");
    let errorContraLong = document.getElementById("errorContraLong");
    let errorContraLong2 = document.getElementById("errorContraLong2");
    let errorCorreoDuplicated = document.getElementById("errorCorreoDuplicated");
    let errorNombre = document.getElementById("errorNombre");
    let errorApellido = document.getElementById("errorApellido");
    let errorCorreo = document.getElementById("errorCorreo");
    let errorTelefono = document.getElementById("errorTelefono");
    let errorTelefonoLong = document.getElementById("errorTelefonoLong");
    let errorContrasenia = document.getElementById("errorContrasenia");
    let errorContrasenia2 = document.getElementById("errorContrasenia2");

    let cedula = document.getElementById("InputCedula");
    let nombre = document.getElementById("InputNombre");
    let apellido = document.getElementById("InputApellido");
    let correo = document.getElementById("InputCorreo");
    let telefono = document.getElementById("InputTelefono");
    let contrasenia = document.getElementById("InputContrasenia");
    let contrasenia2 = document.getElementById("InputContrasenia2");

    if (correo) {
        correo.classList.remove('input-error');
    }
    
    if (cedula) {
        cedula.classList.remove('input-error');
    }

    if (nombre) {
        nombre.classList.remove('input-error');
    }

    if (apellido) {
        apellido.classList.remove('input-error');
    }
    
    if (telefono) {
        telefono.classList.remove('input-error');
    }

    if (errorTelefonoLong) {
        errorTelefonoLong.classList.remove('input-error');
    }

    if (contrasenia) {
        contrasenia.classList.remove('input-error');
    }

    if (contrasenia2) {
        contrasenia2.classList.remove('input-error');
    }

    if (errorCedula) {
        errorCedula.classList.add('oculto');
    }

    if (errorCedulaLong) {
        errorCedulaLong.classList.add('oculto');
    }

    if (errorNombre) {
        errorNombre.classList.add('oculto');
    }

    if (errorApellido) {
        errorApellido.classList.add('oculto');
    }

    if (errorCorreo) {
        errorCorreo.classList.add('oculto');
    }
    
    if (errorTelefono) {
        errorTelefono.classList.add('oculto');
    }
    
    if (errorTelefonoLong) {
        errorTelefonoLong.classList.add('oculto');
    }

    if (errorContrasenia) {
        errorContrasenia.classList.add('oculto');
    }

    if (errorContraLong) {
        errorContraLong.classList.add('oculto');
    }

    if (errorContrasenia2) {
        errorContrasenia2.classList.add('oculto');
    }

    if (errorContraLong2) {
        errorContraLong.classList.add('oculto');
    }

    if (errorCorreoDuplicated) {
        errorCorreoDuplicated.classList.add('oculto');
    }
}

function compararContrasenias() {
    let contrasenia = document.getElementById("InputContrasenia").value;
    let contrasenia2 = document.getElementById("InputContrasenia2").value;
    let errorPwdDesigual = document.getElementById("errorPwdDesigual");


    if (contrasenia === contrasenia2) {
        errorPwdDesigual.classList.add('oculto');
    } else {
        errorPwdDesigual.classList.remove('oculto');
    }

}

File: src/java/Servlets/Registro.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/JSP_Servlet/Servlet.java to edit this template
 */
package Servlets;

import java.io.IOException;
import java.io.PrintWriter;
import java.io.UnsupportedEncodingException;
import java.security.GeneralSecurityException;
import java.security.NoSuchAlgorithmException;
import java.security.spec.InvalidKeySpecException;
import java.util.logging.Level;
import java.util.logging.Logger;
import javax.servlet.RequestDispatcher;
import javax.servlet.ServletException;
import javax.servlet.annotation.WebServlet;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import logica.clases.Cliente;
import logica.clases.Usuario;
import logica.fabrica.Fabrica;
import logica.interfaces.IAdministracion;
import javax.servlet.http.HttpSession;

/**
 *
 * @author gasto
 */
@WebServlet(name = "Registro", urlPatterns = {"/Registro"})
public class Registro extends HttpServlet {

    /**
     * Processes requests for both HTTP <code>GET</code> and <code>POST</code>
     * methods.
     *
     * @param request servlet request
     * @param response servlet response
     * @throws ServletException if a servlet-specific error occurs
     * @throws IOException if an I/O error occurs
     */
    protected void processRequest(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {
        response.setContentType("text/html;charset=UTF-8");
        try (PrintWriter out = response.getWriter()) {
            /* TODO output your page here. You may use following sample code. */
            out.println("<!DOCTYPE html>");
            out.println("<html>");
            out.println("<head>");
            out.println("<title>Servlet Login</title>");
            out.println("</head>");
            out.println("<body>");
            out.println("<h1>Servlet Login at " + request.getContextPath() + "hola</h1>");
            out.println("</body>");
            out.println("</html>");
        }
    }

    // <editor-fold defaultstate="collapsed" desc="HttpServlet methods. Click on the + sign on the left to edit the code.">
    /**
     * Handles the HTTP <code>GET</code> method.
     *
     * @param request servlet request
     * @param response servlet response
     * @throws ServletException if a servlet-specific error occurs
     * @throws IOException if an I/O error occurs
     */
    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {
        RequestDispatcher dispatcher = request.getRequestDispatcher("/Vistas/Registro.jsp");
        dispatcher.forward(request, response);
    }

    /**
     * Handles the HTTP <code>POST</code> method.
     *
     * @param request servlet request
     * @param response servlet response
     * @throws ServletException if a servlet-specific error occurs
     * @throws IOException if an I/O error occurs
     */
    @Override
    protected void doPost(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException, UnsupportedEncodingException {

        IAdministracion IA = Fabrica.getInstancia().getControladorCliente();
        int cedula = Integer.parseInt(request.getParameter("InputCedula"));
        String nombre = request.getParameter("InputNombre");
        String apellido = request.getParameter("InputApellido");
        String correo = request.getParameter("InputCorreo");
        String contrasenia = request.getParameter("InputContrasenia");
        String contrasenia2 = request.getParameter("InputContrasenia2");
        Cliente cliente = IA.traerClienteSeleccionado(cedula);

        Usuario user = IA.obtenerUsuario(correo);
        ProtectedUserPassword encryptedPassword = null;

        try {
            encryptedPassword = new ProtectedUserPassword(contrasenia);
        } catch (NoSuchAlgorithmException ex) {
            Logger.getLogger(Registro.class.getName()).log(Level.SEVERE, null, ex);
        } catch (InvalidKeySpecException ex) {
            Logger.getLogger(Registro.class.getName()).log(Level.SEVERE, null, ex);
        }
        
        String encryptPassword = "";
        try {
            encryptPassword = encryptedPassword.encrypt();
        } catch (GeneralSecurityException ex) {
            Logger.getLogger(Registro.class.getName()).log(Level.SEVERE, null, ex);
        }

        if (!contrasenia.equals(contrasenia2)) {
            request.setAttribute("correo", correo);
            request.setAttribute("nombre", nombre);
            request.setAttribute("cedula", cedula);
            request.setAttribute("apellido", apellido);
            request.setAttribute("contrasenia", contrasenia);
            request.setAttribute("contrasenia2", contrasenia2);
            request.getRequestDispatcher("/Vistas/Registro.jsp").forward(request, response);
        } 
        
        if (user == null) {
            IA.crearUsuario(correo, encryptPassword, encryptedPassword.getKey().getEncoded());
        } else {
            request.setAttribute("correo", correo);
            request.setAttribute("nombre", nombre);
            request.setAttribute("cedula", cedula);
            request.setAttribute("apellido", apellido);
            request.getRequestDispatcher("/Vistas/Registro.jsp").forward(request, response);
        }

        if (cliente != null) {
            IA.editarClienteSeleccionado(cedula, nombre, apellido, 0, correo);
        } else {
            IA.agregarCliente(cedula, nombre, apellido, 0, correo);
        }
        HttpSession session = request.getSession(true);
        session.setAttribute("user", correo);

        response.sendRedirect("/");

    }

    /**
     * Returns a short description of the servlet.
     *
     * @return a String containing servlet description
     */
    @Override
    public String getServletInfo() {
        return "Short description";
    }// </editor-fold>

}
